# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Compute \eqn{F_ST} between two subpopulations.
#' 
#' Calculates Hudson's \eqn{F_ST} across genetic marker(s) for two subpopulations.
#' 
#' @param X1 A \eqn{n_1} x \eqn{m} numeric matrix of allelic dosages for one 
#'   subpopulation.
#' @param X2 A \eqn{n_2} x \eqn{m} numeric matrix of allelic dosages for a 
#'   different subpopulation.
#'   
#' @details The entries of \code{X1} and \code{X2} should be in the interval 
#'   \eqn{[0,2]}. This function implements the recommended Hudson's \eqn{F_ST} 
#'   estimator (equation 5) from the reference. When \eqn{m>1}, the function uses 
#'   the recommended ratio of averages approach for combining estimates of 
#'   \eqn{F_ST} across multiple markers.
#' 
#' @return A single numeric estimate for \eqn{F_ST} between two subpopulations.
#' 
#' @references Bhatia \emph{et al.} (2013) Estimating and interpreting F_ST: The 
#'   impact of rare variants. \emph{Genome Research}, \strong{23}\emph{(9)}: 
#'   1514-21.
#' 
#' @export
hudson_fst <- function(X1, X2) {
    .Call(`_QGenTools_hudson_fst`, X1, X2)
}

#' Identity-by-state between genotypes.
#' 
#' Calculates the proportion of shared alleles between pairs of genotypes.
#' 
#' @param X A numeric matrix of allelic dosages where rows represent different
#'   genotypes and columns represent different markers. If \code{X} has rownames, 
#'   these will be preserved in the output.
#' 
#' @details Each element of \code{X} should be in the interval \eqn{[0,2]}. 
#'   The matrix is pre-multiplied by one-half to correspond to equation 1 of 
#'   Bustos-Korts \emph{et al.} (2016).
#' 
#' @return A numeric matrix containing the proportion of shared alleles between
#'   each pair of genotypes.
#' 
#' @references Bustos-Korts \emph{et al.} (2016) Improvement of predictive 
#'   ability by uniform coverage of the target genetic space. \emph{G3}, 
#'   \strong{6}: 3733-47.
#' 
#' @export
ibs <- function(X) {
    .Call(`_QGenTools_ibs`, X)
}

#' Correlation between loci.
#' 
#' Calculates the correlation (\eqn{r^2}) between marker loci within a population.
#' 
#' @param X An \eqn{n} x \eqn{m} numeric matrix of allelic dosages. The matrix must 
#'   have column names for labelling correlations in the output.
#' 
#' @details Elements of \code{X} should be in the interval \eqn{[0,2]}. This 
#'   computes the traditional \eqn{r^2} measure of linkage disequilibrium 
#'   between pairs of marker loci.
#' 
#' @return A dataframe with \eqn{m(m-1)/2} rows containing pairwise correlations
#'   between all possible pairs of markers in \code{X}.
#' 
#' @export
r2 <- function(X) {
    .Call(`_QGenTools_r2`, X)
}

#' Correlation between loci.
#' 
#' Calculates the structure corrected correlation (\eqn{r_S^2}) between marker 
#'   loci within a population.
#' 
#' @param X An \eqn{n} x \eqn{m} numeric matrix of allelic dosages. The matrix must 
#'   have column names for labelling correlations in the output.
#' @param st An \eqn{n} x \eqn{K-1} numeric matrix of population structure 
#'   covariates where \eqn{K} is the estimated number of subpopulations.
#' 
#' @details Elements of \code{X} should be in the interval \eqn{[0,2]}. \code{st} 
#'   is usually produced by principal components analysis or contains admixture 
#'   proportions from a program like STRUCTURE or fastSTRUCTURE. This function 
#'   calculates the structure corrected correlation (\eqn{r_S^2}) between marker
#'   loci (equation 1 of the reference).
#' 
#' @return A dataframe with \eqn{m(m-1)/2} rows containing pairwise correlations
#'   between all possible pairs of markers in \code{X}.
#'
#' @references Mangin \emph{et al.} (2011) Novel measures of linkage disequilibrium
#'   that correct the bias due to population structure and relatedness. 
#'   \emph{Heredity}, \strong{108}\emph{(3)}: 285-91.
#'
#' @export
r2s <- function(X, st) {
    .Call(`_QGenTools_r2s`, X, st)
}

#' Correlation between loci.
#' 
#' Calculates the kinship corrected correlation (\eqn{r_V^2}) between marker 
#'   loci within a population.
#' 
#' @param X An \eqn{n} x \eqn{m} numeric matrix of allelic dosages. The matrix must 
#'   have column names for labelling correlations in the output.
#' @param V An \eqn{n} x \eqn{n} numeric matrix of realized genetic relationships.
#' 
#' @details Elements of \code{X} should be in the interval \eqn{[0,2]}. \code{r2v} 
#'   computes the Moore-Penrose pseudoinverse of \code{V} on demand, so singular 
#'   \code{V} are permitted. This function calculates the kinship corrected 
#'   correlation (\eqn{r_V^2}) between marker loci (equation 2 of the reference).
#' 
#' @return A dataframe with \eqn{m(m-1)/2} rows containing pairwise correlations
#'   between all possible pairs of markers in \code{X}.
#'
#' @references Mangin \emph{et al.} (2011) Novel measures of linkage disequilibrium
#'   that correct the bias due to population structure and relatedness. 
#'   \emph{Heredity}, \strong{108}\emph{(3)}: 285-91.
#'
#' @export
r2v <- function(X, V) {
    .Call(`_QGenTools_r2v`, X, V)
}

#' Correlation between loci.
#' 
#' Calculates the kinship and structure corrected correlation (\eqn{r_VS^2}) 
#'   between marker loci within a population.
#' 
#' @param X An \eqn{n} x \eqn{m} numeric matrix of allelic dosages. The matrix must 
#'   have column names for labelling correlations in the output.
#' @param V An \eqn{n} x \eqn{n} numeric matrix of realized genetic relationships.
#' @param st An \eqn{n} x \eqn{K-1} numeric matrix of population structure 
#'   covariates where \eqn{K} is the estimated number of subpopulations.
#' 
#' @details Elements of \code{X} should be in the interval \eqn{[0,2]}. \code{r2v} 
#'   computes the Moore-Penrose pseudoinverse of \code{V} on demand, so singular 
#'   \code{V} are permitted. \code{st} is usually produced by principal components 
#'   analysis or contains admixture proportions from a program like STRUCTURE or 
#'   fastSTRUCTURE. This function calculates the kinship and structure corrected 
#'   correlation (\eqn{r_VS^2}) between marker loci (equation 3 of the reference).
#' 
#' @return A dataframe with \eqn{m(m-1)/2} rows containing pairwise correlations
#'   between all possible pairs of markers in \code{X}.
#'
#' @references Mangin \emph{et al.} (2011) Novel measures of linkage disequilibrium
#'   that correct the bias due to population structure and relatedness. 
#'   \emph{Heredity}, \strong{108}\emph{(3)}: 285-91.
#'
#' @export
r2vs <- function(X, V, st) {
    .Call(`_QGenTools_r2vs`, X, V, st)
}

#' Realized genetic relationships between genotypes.
#' 
#' Calculates pairwise genetic relationships based on marker scores.
#' 
#' @param X A numeric matrix of allelic dosages. The rows should correspond to 
#'   genotypes and the columns should correspond to markers. If \code{X} has 
#'   rownames, these will be preserved in the output.
#' 
#' @details Each element of \code{X} should be in the interval \eqn{[0,2]}.
#' 
#' @return The matrix of realized genetic relationships (variances and covariances) 
#'   between each pair of genotypes.
#' 
#' @references Astle and Balding. (2009) Population structure and cryptic 
#'   relatedness in genetic association studies. \emph{Statistical Science}, 
#'   \strong{24}\emph{(4)}: 451-71.
#' 
#' @export
realized_ab <- function(X) {
    .Call(`_QGenTools_realized_ab`, X)
}

